contracts/attesters/hydra-s1/HydraS1AccountboundAttester.sol

```diff
@@ -0,0 +1,258 @@
+   * @param collectionIdLast Id of the last collection in which the attester is supposed to record
+   */
+  constructor(
+    address attestationsRegistryAddress,
+    address hydraS1VerifierAddress,
+    address availableRootsRegistryAddress,
+    address commitmentMapperAddress,
+    uint256 collectionIdFirst,
+    uint256 collectionIdLast
+  )
+    HydraS1SimpleAttester(
+      attestationsRegistryAddress,
+      hydraS1VerifierAddress,
+      availableRootsRegistryAddress,
+      commitmentMapperAddress,
+      collectionIdFirst,
+      collectionIdLast
+    )
+  {}
+
+  /*******************************************************
+    MANDATORY FUNCTIONS TO OVERRIDE FROM ATTESTER.SOL
+  *******************************************************/
+
+  /**
+   * @dev Returns the actual attestations constructed from the user request
+   * @param request users request. Claim of having an account part of a group of accounts
+   * @param proofData snark public input as well as snark proof
+   */
+  function buildAttestations(Request calldata request, bytes calldata proofData)
+    public
+    view
+    virtual
+    override(IAttester, HydraS1SimpleAttester)
+    returns (Attestation[] memory)
+  {
+    Attestation[] memory attestations = super.buildAttestations(request, proofData);
+
+    uint256 nullifier = proofData._getNullifier();
+    attestations[0].extraData = abi.encodePacked(
+      attestations[0].extraData,
+      encodeNullifierAndNullifierBurnCount(nullifier, attestations[0].owner)
+    );
+
+    return (attestations);
+  }
+
+  /*******************************************************
+    OPTIONAL HOOK VIRTUAL FUNCTIONS FROM ATTESTER.SOL
+  *******************************************************/
+  /**
+   * @dev Hook run before recording the attestation.
+   * Throws if nullifier already used, not a renewal, and nullifier on cooldown.
+   * @param request users request. Claim of having an account part of a group of accounts
+   * @param proofData provided to back the request. snark input and snark proof
+   */
+  function _beforeRecordAttestations(Request calldata request, bytes calldata proofData)
+    internal
+    virtual
+    override
+  {
+    uint256 nullifier = proofData._getNullifier();
+    address nullifierDestination = _getDestinationOfNullifier(nullifier);
+
+    HydraS1Claim memory claim = request._claim();
+
+    // check if the nullifier has already been used previously, if so it may be on cooldown
+    if (nullifierDestination != address(0) && nullifierDestination != claim.destination) {
+      uint32 cooldownDuration = _getCooldownDurationForGroupId(claim.groupId);
+      if (_isOnCooldown(nullifier, cooldownDuration)) {
+        uint16 burnCount = _getNullifierBurnCount(nullifier);
+        revert NullifierOnCooldown(nullifier, nullifierDestination, burnCount, cooldownDuration);
+      }
+
+      // Delete the old Attestation on the account before recording the new one
```

Comment:
Move all the "delete attestation" logic in an other function?
The `beforeRecordAttestations` would be clearer